certstore/certstore.go

```diff
@@ -121,44 +246,156 @@ func (cs *Store) GetRange(ctx context.Context, start uint64, end uint64) ([]cert
 	return certs, nil
 }
 
-func (*Store) keyForInstance(i uint64) datastore.Key {
+func (cs *Store) readPowerTable(ctx context.Context, instance uint64) (gpbft.PowerEntries, error) {
+	var powerTable gpbft.PowerEntries
+	if b, err := cs.ds.Get(ctx, cs.keyForPowerTable(instance)); err != nil {
+		return nil, xerrors.Errorf("failed to load power table at instance %d: %w", instance, err)
+	} else if err := powerTable.UnmarshalCBOR(bytes.NewReader(b)); err != nil {
+		return nil, xerrors.Errorf("failed to unmarshal power table at instance %d: %w", instance, err)
+	}
+	return powerTable, nil
+}
+
+// Store the specified power table.
+func (cs *Store) putPowerTable(ctx context.Context, instance uint64, powerTable gpbft.PowerEntries) error {
+	var buf bytes.Buffer
+	if err := powerTable.MarshalCBOR(&buf); err != nil {
+		return xerrors.Errorf("marshalling power table instance %d: %w", instance, err)
+	}
+	if err := cs.ds.Put(ctx, cs.keyForPowerTable(instance), buf.Bytes()); err != nil {
+		return xerrors.Errorf("putting power table instance %d: %w", instance, err)
+	}
+	return nil
+}
+
+// GetPowerTable returns the power table (committee) used to validate the specified instance.
+func (cs *Store) GetPowerTable(ctx context.Context, instance uint64) (gpbft.PowerEntries, error) {
+	if instance < cs.firstInstance {
+		return nil, fmt.Errorf("cannot return a power table before the first instance: %d", cs.firstInstance)
+	}
+
+	lastPowerTable := cs.firstInstance
+	if latestCert := cs.Latest(); latestCert != nil {
+		lastPowerTable = latestCert.GPBFTInstance + 1
+	}
+	if instance > lastPowerTable {
+		return nil, fmt.Errorf("cannot return future power table for instance %d > %d", instance, lastPowerTable)
+	}
+
+	// We store every `powerTableFrequency` power tables. Find the nearest multiple smaller than
+	// the requested instance.
+	startInstance := instance
+	if offset := instance % cs.powerTableFrequency; offset != 0 {
+		startInstance = max(startInstance-offset, cs.firstInstance)
```

Comment:
IMO, this doesn't have to be in an if, and I had to think about it for a second to check that skipping the `max` is safe.

Reply:
We still need the max as the first instance may not be aligned. But you're right, we don't need the if.

Reply:
Fixed.